code/extensions/che-commands/src/utils.ts

```diff
@@ -0,0 +1,40 @@
+/**********************************************************************
+ * Copyright (c) 2023 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+/* eslint-disable header/header */
+
+import { V1alpha2DevWorkspaceSpecTemplateCommandsItemsExecEnv } from "@devfile/api";
+
+/*
+ * This function is used to resolve env variables in command line.
+ * It should support both syntaxes: ${VAR} and $VAR
+ * The function returns the original command line if:
+ * - envVariables is undefined or empty
+ * - there are no env variables in the command line
+ * - there are env variables in the command line but they are not defined in the envVariables array
+ */
+export function resolveEnvVariablesForCommand(commandLine: string, envVariables?: Array<V1alpha2DevWorkspaceSpecTemplateCommandsItemsExecEnv>): string {
+	if (!envVariables || envVariables.length === 0) {
+		return commandLine;
+	}
+
+	// it should support both syntaxes: ${VAR} and $VAR
+	const regex = /\$\{?([a-zA-Z_][a-zA-Z0-9_]*)\}?/g;;
+	let match = regex.exec(commandLine);
+	while (match) {
+		const envVariableName = match[1];
+		const envVariableValue = envVariables.find(env => env.name === envVariableName)?.value;
```

Comment:
`envVariables` array contains only the variables defined in the devfile. It doesn't read from `process.env`, correct?
I didn't get how resolving the "real" environment variables works (like `KUBERNETES_SERVICE_HOST`).
Could you point me out, please?

Reply:
let's say a command looks like:
`echo "KUBERNETES_SERVICE_HOST ${KUBERNETES_SERVICE_HOST}"`

for such use case function `resolveEnvVariablesForCommand()` should return the command as is:
`echo "KUBERNETES_SERVICE_HOST ${KUBERNETES_SERVICE_HOST}"`

the env variable is resolved directly in the terminal if it's defined.
try to execute the same command in a terminal for the `tools` container: 
<img width="504" alt="image" src="https://user-images.githubusercontent.com/5676062/234250686-81410f7a-bf1d-45e2-a70e-39af8cf00cd8.png">

so, there is no magic  :-)

Reply:
Got it. Thanks!